### DISCUSSION TITLE
Questions and feedback for tab-switching userscript

### DISCUSSION THREAD

**Discussion by AckslD** [qutebrowser/qutebrowser#5959] [Q&A] [answered]
I made an attempt to make a userscript to quickly switch tabs using `rofi`, which can be found at https://github.com/AckslD/qute-switch-tab.

I'm curious if something similar already exists and have some questions about the way I implemented it and if there is a more "qute-ic" way. Feel free to close this if this is not the right medium for such a discussion :)

Concretely I have the following questions:
* To find what tabs are currently open, the current session needs to be saved and parsed. Can this be done in some other way?
* Is there a way to avoid the error `No command given` if one does not make a choice and cancels the command?
* How to avoid the bash-file `switch_tab`? I first only had the python file and tried to do
  ```python
  os.environ["QUTE_FIFO"] += f"tab-focus {tab_num + 1}"
  ```
  but the change to the environment variable didn't seem to have an effect. This is why the bash-file `switch_tab` just wraps `switch_tab.py` to set the environment variable.
* Currently `switch_tab` can only switch between tabs in the same window, how could one switch between all tabs? Probably needs to interact with the window/tiling manager?
* To know which tab is chosen, the information about the tab number is show in the choices passed to `rofi`. Not a big issue but maybe some way to avoid this?

**Comment by toofar** [comment]
You should be writing to the fifo from python like this https://github.com/qutebrowser/qutebrowser/blob/master/misc/userscripts/ripbang#L32

> how could one switch between all tabs?

The :buffer command can take an argument like `{win_num}/{tab_num}`, both starting at 1. It'll raise the window it is different than the current one. It also already has a builtin completion in case you hadn't seen that already.

**Comment by AckslD** [comment]
Oh that's cool :) The builtin completion of `:buffer` is what I wanted haha. The only thing I'd be missing is to avoid typing `:buffer`, is there a way to have a keybinding that fills the command-line without executing the command? So that one can press for example ctrl-p and then start typing something that would match a tab.

**Comment by The-Compiler** [comment, marked as answer]
`:set-cmd-text -s :buffer`, which is what `gt` is already bound to by default.

Also I'm converting this to a discussion, as this isn't an issue with qutebrowser.

**Reply by AckslD** [reply]
@The-Compiler That's perfect! Should have researched this a bit more beforehand haha. It was nevertheless a fun exercise to create the userscript :)
